services/web/server/src/simcore_service_webserver/projects/projects_db.py

```diff
@@ -239,55 +140,74 @@ async def add_project(
+                primary_gid = await self._get_user_primary_group_gid(
+                    conn, user_id=user_id
+                )
+                project_db_values.setdefault("access_rights", {}).update(
+                    create_project_access_rights(primary_gid, ProjectAccessRights.OWNER)
                 )
             # ensure we have the minimal amount of data here
-            kargs.setdefault("name", "New Study")
-            kargs.setdefault("description", "")
-            kargs.setdefault("workbench", {})
+            project_db_values.setdefault("name", "New Study")
+            project_db_values.setdefault("description", "")
+            project_db_values.setdefault("workbench", {})
             # must be valid uuid
             try:
-                uuidlib.UUID(str(kargs.get("uuid")))
+                uuidlib.UUID(str(project_db_values.get("uuid")))
             except ValueError:
                 if force_project_uuid:
                     raise
-                kargs["uuid"] = str(uuidlib.uuid1())
-
-            # insert project
-            retry = True
-            while retry:
-                try:
-                    query = projects.insert().values(**kargs)
-                    await conn.execute(query)
-                    retry = False
-                except psycopg2.errors.UniqueViolation as err:  # pylint: disable=no-member
-                    if (
-                        err.diag.constraint_name != "projects_uuid_key"
-                        or force_project_uuid
-                    ):
-                        raise
-                    kargs["uuid"] = str(uuidlib.uuid1())
-                    retry = True
-
-            # insert projects_to_product entry
-            await self.upsert_project_linked_product(
-                ProjectID(f"{kargs['uuid']}"), product_name, conn=conn
-            )
+                project_db_values["uuid"] = f"{uuidlib.uuid1()}"
+
+            async with conn.begin() as transaction:
+                # atomic transaction to insert project and update relations
+                project_index = None
+                retry = True
+                while retry:
+                    try:
+                        project_index = await conn.scalar(
+                            projects.insert()
+                            .values(**project_db_values)
+                            .returning(projects.c.id)
+                        )
+                        retry = False
+                    except UniqueViolation as err:
+                        if (
+                            err.diag.constraint_name != "projects_uuid_key"
+                            or force_project_uuid
+                        ):
+                            raise
+                        # tries new uuid, but before rolls back
+                        # previous execution from transaction since it should
+                        # not execute at the end
+                        await transaction.rollback()
```

Comment:
are you sure that this rollback is necessary? if the insertion failed, I think it should not be inserted?

Reply:
Yes, i think it is necessary. The rollback is to remove the failed execution from the transactions. If I do not do this, it throws an exception telling me exactly that

Reply:
Changed as agreed: using tenacity on the transaction context